cpp/src/join/join.cu

```diff
@@ -0,0 +1,310 @@
+  }
+
+  // If Inner Join and either table is empty, return immediately
+  if ((join_type::INNER_JOIN == join_t) &&
+      ((0 == left.num_rows()) || (0 == right.num_rows()))) {
+      return true;
+  }
+
+  return false;
+}
+
+template <typename index_type>
+std::unique_ptr<experimental::table>
+get_trivial_full_join_indices(
+    table_view const& left,
+    table_view const& right,
+    cudaStream_t stream,
+    rmm::mr::device_memory_resource* mr) {
+
+  CUDF_EXPECTS((left.num_rows() != 0) || (right.num_rows() != 0),
+      "Expecting at least one table view to have non-zero number of rows");
+
+  rmm::device_buffer left_index, right_index;
+  size_type result_size = 0;
+
+  if (left.num_rows() == 0) {
+    result_size = right.num_rows();
+    left_index  = rmm::device_buffer{sizeof(index_type)*result_size, stream, mr};
+    thrust::device_ptr<index_type> l(static_cast<index_type*>(left_index.data()));
+    thrust::fill(rmm::exec_policy(stream)->on(stream), l, l + result_size,
+        static_cast<index_type>(-1));
+
+    right_index = rmm::device_buffer{sizeof(index_type)*result_size, stream, mr};
+    thrust::device_ptr<index_type> r(static_cast<index_type*>(right_index.data()));
+    thrust::sequence(rmm::exec_policy(stream)->on(stream), r, r + result_size);
+  } else if (right.num_rows() == 0) {
+    result_size = left.num_rows();
+    left_index  = rmm::device_buffer{sizeof(index_type)*result_size, stream, mr};
+    thrust::device_ptr<index_type> l(static_cast<index_type*>(left_index.data()));
+    thrust::sequence(rmm::exec_policy(stream)->on(stream), l, l + result_size);
+
+    right_index = rmm::device_buffer{sizeof(index_type)*result_size, stream, mr};
+    thrust::device_ptr<index_type> r(static_cast<index_type*>(right_index.data()));
+    thrust::fill(rmm::exec_policy(stream)->on(stream), r, r + result_size,
+        static_cast<index_type>(-1));
+  }
+
+  return get_indices_table<index_type>(
+      std::move(left_index), std::move(right_index),
+      result_size, stream, mr);
+}
+
+template <join_type join_t, typename index_type>
+std::unique_ptr<experimental::table>
+get_base_join_indices(
+    table_view const& left,
+    table_view const& right,
+    cudaStream_t stream,
+    rmm::mr::device_memory_resource* mr) {
+  CUDF_EXPECTS (0 != left.num_columns(), "Selected left dataset is empty");
+  CUDF_EXPECTS (0 != right.num_columns(), "Selected right dataset is empty");
+  CUDF_EXPECTS(std::equal(
+      std::cbegin(left), std::cend(left),
+      std::cbegin(right), std::cend(right),
+      [](const auto &l, const auto &r) {
+      return l.type() == r.type(); }),
+      "Mismatch in joining column data types");
+
+  constexpr join_type base_join_t = (join_t == join_type::FULL_JOIN)? join_type::LEFT_JOIN : join_t;
+  return get_base_hash_join_indices<base_join_t, index_type>(left, right, false, stream, mr);
+}
+
+std::vector<std::unique_ptr<column>>
+combine_join_columns(
+    std::unique_ptr<experimental::table> left_table,
```

Comment:
I'd do it like this and avoid the extra complication of a `unique_ptr`. Just require the caller to pass ownership of the `table` to this function by accepting via r-value ref. 
```suggestion
    experimental::table&& left_table,
```